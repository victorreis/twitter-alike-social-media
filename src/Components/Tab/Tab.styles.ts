import styled, { CSSObject, DefaultTheme } from 'styled-components';

import { TabStyleProps } from './Tab.types';

const getTabActiveBorder = ({
  theme,
  active,
}: {
  theme: DefaultTheme;
  active: boolean;
}): CSSObject => {
  return {
    ...(active && {
      borderBottomWidth: '4px',
      borderBottomStyle: 'solid',
      borderBottomColor: theme.colors.background.default.light,
    }),
  };
};

export const TabContainer = styled.div<TabStyleProps>`
  box-sizing: border-box;
  height: 3rem;
  padding: 16px;
  padding-bottom: 12px;

  color: ${({ theme }) => theme.colors.font.default};
  ${({ theme }) => theme.typographies.button};

  ${getTabActiveBorder};

  &:hover {
    background-color: ${({ theme }) => theme.colors.background.default.dark};
  }
`;
