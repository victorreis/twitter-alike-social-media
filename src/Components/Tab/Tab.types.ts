import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredTabProps {
  /**
   * Tab text.
   */
  text: string;
}

export interface DefaultTabProps {
  /**
   * Controls when the tab is active by showing a visual feedback.
   * @default 'false'.
   */
  active?: boolean;
}

export interface OptionalTabProps {}

export type TabProps = RequiredTabProps &
  DefaultTabProps &
  OptionalTabProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;

export type TabStyleProps = Required<DefaultTabProps>;
