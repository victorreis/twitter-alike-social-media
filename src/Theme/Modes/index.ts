export { darkTheme } from './Dark.theme';
