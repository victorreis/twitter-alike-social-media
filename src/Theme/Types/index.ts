export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
