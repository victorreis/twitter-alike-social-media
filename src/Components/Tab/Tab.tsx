import { TestProps } from '../../Config/Tests/Test.types';
import { TabContainer } from './Tab.styles';
import { TabProps, DefaultTabProps } from './Tab.types';

export const tabDefaults: Required<DefaultTabProps> & Required<TestProps> = {
  testID: 'Tab',
  active: false,
};

export const Tab: React.FC<TabProps> = (props): JSX.Element => {
  const {
    testID = tabDefaults.testID,
    text,
    active = tabDefaults.active,
    style,
    ...others
  } = props;

  return (
    <TabContainer
      active={active}
      data-testid={testID}
      style={style}
      {...others}
    >
      {text}
    </TabContainer>
  );
};
