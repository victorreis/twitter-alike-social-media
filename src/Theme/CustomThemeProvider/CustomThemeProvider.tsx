/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
