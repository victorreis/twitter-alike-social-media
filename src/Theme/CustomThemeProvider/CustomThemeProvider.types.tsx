import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
