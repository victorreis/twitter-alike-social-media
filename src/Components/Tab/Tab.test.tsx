import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb } from '../../Utils/Transform/hexToRgb.util';
import { Tab, tabDefaults } from './Tab';
import { RequiredTabProps, TabProps } from './Tab.types';

describe('tab component tests', () => {
  const text = 'text';
  const newActive = true;

  const requiredProps: RequiredTabProps = {
    text,
  };

  const setup = (props?: TabProps) => {
    const renderRTR = () =>
      renderRTRCreator<TabProps>(Tab, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<TabProps>(Tab, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(tabDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(text);

      expect(element).toBeInTheDocument();
    });

    it(`should render '${String(
      tabDefaults.active
    )}' as the default active status`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        active: tabDefaults.active,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'active' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        active: newActive,
      }).renderRTR().root;
      const element = instance.findByProps({ active: newActive });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(tabDefaults.testID);

      expect(container).toHaveStyle({
        color: hexToRgb(themes.default.colors.font.default),
        ...themes.default.typographies.button,
      });
    });

    it(`should add border style when the tab is active`, () => {
      expect.assertions(1);
      setup({ ...requiredProps, active: true }).renderJestDom();
      const container = screen.getByTestId(tabDefaults.testID);

      expect(container).toHaveStyle({
        borderBottomWidth: '4px',
        borderBottomStyle: 'solid',
        borderBottomColor: themes.default.colors.background.default.light,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
