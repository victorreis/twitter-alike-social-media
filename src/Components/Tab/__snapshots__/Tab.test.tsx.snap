// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tab component tests snapshot tests should render correctly 1`] = `
<div
  className="TabContainer-twitter-alike-social-media__sc-15n3a25-0 cfmzd"
  data-testid="Tab"
>
  text
</div>
`;
